2024/YoctoProjectDeveloperDay/practical-sec-tee.md

```diff
@@ -0,0 +1,994 @@
+          your environment, are very outdated, and lack support;
+        - You do not have source code, so, debugging becames nearly impossible;
+        - If you have a small contract or do not have direct contact with
+          vendor, you will have to wait for support or, if you do not have time,
+          you will have to go for an endless web-digging.
+
+---
+
+# U-Boot integration
+
+1. Turn on `CONFIG_TEE` and `CONFIG_OPTEE` (and related configs, if needed)
+  configs.
+2. Link ATF (`BL31`):
+
+    ```bb
+    EXTRA_OEMAKE:append = " BL31=${DEPLOY_DIR_IMAGE}/atf.elf"
+    ```
+
+3. Link OP-TEE OS:
+
+    ```bb
+    EXTRA_OEMAKE:append = " TEE=${DEPLOY_DIR_IMAGE}/optee/tee.elf"
+    ```
+
+4. Check `binman` configuration, in case you are using other file formats (e.g.
+ raw binary).
+
+???
+
+- Time for this slide: TODO
+- Idea/goal of this slide: A list of steps on how to integrate ATF and OP-TEE
+  into bootloader.
+- What to say:
+    - TODO
+- Notes:
+    - Present some U-Boot configs;
+    - Present linking ATF and OP-TEE binaries;
+    - Present linkinng binaries vs ELF using Binman.
+    - Present `binman`-related problems.
+
+---
+
+# Linux integration
+
+Turn on `CONFIG_TEE` and `CONFIG_OPTEE` (and related, if needed) configs, and
+add OP-TEE OS and its memory to DTS (here are `CFG_TZDRAM_START`,
+`CFG_TZDRAM_SIZE`, `CFG_SHMEM_START` and `CFG_SHMEM_SIZE`):
+
+```dts
+    reserved-memory {
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
+
+		optee@8400000 {
+			reg = <0x0 0x08400000 0x0 0x02000000>;
+            no-map;
+		};
+
+		optee_shm: optee_shared_mem@a400000 {
+			reg = <0x0 0x0a400000 0x0 0x00400000>;
+		};
+	};
+
+	firmware {
+		optee: optee {
+			compatible = "linaro,optee-tz";
+			method = "smc";
+			shm = <&optee_shm>;
+		};
+    };
+```
+
+???
+
+- Time for this slide: TODO
+- Idea/goal of this slide: A list of steps on how to integrate ATF and OP-TEE
+  into Linux.
+- What to say:
+    - TODO
+- Notes:
+    - Present some Linux Kernel configs;
+    - Present devicetree modifications;
+    - Some userspace tools and libs integration.
+
+---
+
+# Adding packages and recipes:
+
+1. Add TFA `.bbappend` with configuration according to
+  `meta-arm/recipes-bsp/trusted-firmware-a/trusted-firmware-a_2.10.3.bb`;
+2. Add OP-TEE OS `.bbappend` with configuration according to
+  `meta-arm-bsp/recipes-bsp/trusted-firmware-a/trusted-firmware-a_%.bbappend`;
+3. You can add OP-TEE OS testsuite according to
+  `meta-arm-bsp/recipes-security/optee/optee-test_4.%.bbappend`;
+4. Add the packages to your packagegroup:
+
+    ```bb
+    RDEPENDS:${PN}-security = " \
+        trusted-firmware-a \
+        optee-client \
+        optee-os \
+        optee-test \
+        libp11 \
+    "
+    ```
+???
+
+- Time for this slide: TODO
+- Idea/goal of this slide: TODO
+- What to say:
+    - TODO
+- Notes:
+    - TODO
+
+---
+
+# OP-TEE Secure Storage
+
+.center[ <img src="./img/op-tee-secure-storage.svg" height="350px"> ]
```

Comment:
You haven't mentioned how Secure Storage is implemented on the device. Did you provision the RPMB somehow? Were there any constraints? What are requirements for such implementation? Could you do it differently?

Reply:
> Could you do it differently?

How exactly?

Reply:
On what device? This slide shows Secure Storage implementation in OP-TEE not on any device.

Reply:
What do you understand by Secure Storage? Are the secrets stored on filesystem but are encrypted? Where are then decryption keys? They could be on eMMC without or with RPMB, they could even be fetched from a cloud (and then theres a huge rabbit hole on how to do that securely). So please elaborate what we did in fact implement, why and how as it seems it could be most interesting to the atendees. Also how does OP-TEE interact with the "backend" of secure storage.

Reply:
I am doing it right now, be patient.

Reply:
@tym2k1, I have added notes for all my slides here 9f22666beaedbfd2c31973b413bc9a09c2b81f3b. Check them out and say if you have any questions.

About our effort in integrating all of this. I see I have some spare time in my part (around 3 minutes), so I would be able to provide more details, and I am going to do this now.